File: Searching Algorithms/Binary Search/Python/binarySearch.py
#binary search algorithm through recurssion

def binarySearch(array, target, left_index, right_index):
    middle_index = (left_index + right_index)//2
    if target == array[middle_index]:
        return middle_index

    if right_index < left_index:
        return -1

    if target < array[middle_index]:
        right_index = middle_index - 1
        return binarySearch(array, target, left_index, right_index)

    if target > array[middle_index]:
        left_index = middle_index + 1
        return binarySearch(array, target, left_index, right_index)


numbers = [1,3,5,7,8,9,12,34,35,57,78]
left_index = 0
right_index = len(numbers) - 1
target = int(input("Enter the target value: "))
index = binarySearch(numbers,target,left_index,right_index)
if index == -1:
    print("The target value is not in the list.")
else:
    print("The index of the target value is ", index, " .")

File: Searching Algorithms/Binary Search/Java/src/BinarySearch.java
public class BinarySearch {
    // binary search can be implemented using recursion and iteration. Implementation for recurssion is below.

    //Recursive Implementation
    public int binarySearchRec(int[] array, int target, int left, int right) {
        int middle = (left + right) / 2;

        if (array[middle] == target) {
            return middle;
        }

        if (right < left){
            return -1;
        }

        if (target < array[middle]) {
            return binarySearchRec(array, target, left, middle - 1);
        }

        return binarySearchRec(array, target, middle + 1, right);
    }
}
